python/cudf/cudf/_lib/pylibcudf/scalar.pyx

```diff
@@ -0,0 +1,119 @@
+
+    def __init__(self, pyarrow.lib.Scalar value=None):
+        # TODO: This case is not something we really want to
+        # support, but it here for now to ease the transition of
+        # DeviceScalar.
+        if value is not None:
+            raise ValueError("Scalar should be constructed with a factory")
+
+    @staticmethod
+    def from_pyarrow_scalar(pyarrow.lib.Scalar value, DataType data_type=None):
+        # Allow passing a dtype, but only for the purpose of decimals for now
+
+        # Need a local import here to avoid a circular dependency because
+        # from_arrow_scalar returns a Scalar.
+        from .interop import from_arrow_scalar
+
+        cdef Scalar s = from_arrow_scalar(value)
+        if s.type().id() != type_id.DECIMAL128:
+            if data_type is not None:
+                raise ValueError(
+                    "dtype may not be passed for non-decimal types"
+                )
+            return s
+
+        if data_type is None:
+            raise ValueError(
+                "Decimal scalars must be constructed with a dtype"
+            )
+
+        cdef type_id tid = data_type.id()
+        if tid not in (type_id.DECIMAL32, type_id.DECIMAL64, type_id.DECIMAL128):
+            raise ValueError(
+                "Decimal scalars may only be cast to decimals"
+            )
+
+        if tid == type_id.DECIMAL128:
+            return s
+
+        if tid == type_id.DECIMAL32:
+            s.c_obj.reset(
+                new fixed_point_scalar[decimal32](
+                    (<fixed_point_scalar[decimal128]*> s.c_obj.get()).value(),
+                    scale_type(-value.type.scale),
+                    s.c_obj.get().is_valid()
+                )
+            )
+        elif tid == type_id.DECIMAL64:
+            s.c_obj.reset(
+                new fixed_point_scalar[decimal64](
+                    (<fixed_point_scalar[decimal128]*> s.c_obj.get()).value(),
+                    scale_type(-value.type.scale),
+                    s.c_obj.get().is_valid()
+                )
+            )
+        return s
+
+    cpdef to_pyarrow_scalar(self, metadata):
+        from .interop import to_arrow_scalar
+        return to_arrow_scalar(self, metadata)
+
+    cdef const scalar* get(self) except *:
+        return self.c_obj.get()
+
+    cpdef DataType type(self):
+        """The type of data in the column."""
+        return self._data_type
+
+    cpdef bool is_valid(self):
+        """True if the scalar is valid, false if not"""
+        return self.get().is_valid()
+
+    @staticmethod
+    cdef Scalar from_libcudf(unique_ptr[scalar] libcudf_scalar, dtype=None):
+        """Construct a Scalar object from a libcudf scalar.
+
+        This method is for pylibcudf's functions to use to ingest outputs of
+        calling libcudf algorithms, and should generally not be needed by users
+        (even direct pylibcudf Cython users).
+        """
+        cdef Scalar s = Scalar.__new__(Scalar)
```

Comment:
That's a great find. I think you are correct. Unfortunately I don't know if we have a great way to solve this problem right now. rmm DeviceMemoryResource objects can't be easily constructed from librmm objects AFAIK, so we can't extract the one from the underlying scalar to create a suitable Python object to store to the pylibcudf.Scalar instance. The way I see it we have two options:
1. Implement that sort of constructor in some way
2. Modify librmm scalars to keep their memory resources alive via a shared pointer or so

2 is a very far-reaching change that I don't see happening anytime soon (I assume there was a reason for the current design where rmm::device_buffer doesn't share ownership of the allocating mr, but we may want to revisit that design), so we may have to wait on 1 to fix this. Note that this isn't a new issue for cudf, just something we're noticing now.

@shwina WDYT?

Reply:
>  so we may have to wait on 1 to fix this.

I don't  think being able to construct a `DeviceMemoryResource` from a librmm `device_memory_resource *` (which is what `device_buffer` and `device_scalar` carry around) would work. Since on the librmm side we just have a raw pointer, we have no means of keeping it alive. AIUI, the RMM philosophy is that memory resources and device_buffers are RAII and it is up to the user code to manage the lifetime of the memory resource attached to any given device buffer.

I think this means that if you have some object:

```c++
struct foo {
    ...
    rmm::device_buffer<...> _data;
}
```
and your library creates objects of type `foo` with unique ownership semantics (i.e. it hands back `std::unique_ptr<foo>` from creation routines), there isn't a safe way to take ownership of that object unless the library _also_ provides a way of transferring (or sharing) ownership of the `memory_resource` it used when making the `foo` things.

Reply:
You're right, but let's continue discussion of mrs on the other open thread.